Improve the performance of source_bytes computation during gpexpand.

kainwen opened the pull request:
During gpexpand, it will record all the tables that need to do data
redistribution and in order to show the progress the table size is
also saved in the status_details table.

Previously, we directly invoke pg_relation_size to compute that value
in the target list of a SQL involving only catalog tables. This will
lead to very bad performance because for each table it will do a dispatch
to compute its size and the whole progress is done in serial.

This commit rewrite the SQL generating the status_details information.
The skills used to MPP compute table size is:
  1. we use gp_dist_random('pg_class') to get a force random RangeTable
  2. since pg_relations_size is volatile, wrap the computation in a
     subquery before group-by to prevent optimizer to pull up subquery
     and force the pg_relation_size is evaluated before motion then
     sum it together
  3. join with the main SQL

We add test cases in regress/subselect_gp and regress/db_size_functions
to guard the above SQL skill.

We have done local test to show the performance gain. In a local 8*8
64-segments cluster, we create 100 partition tables, each partition
table contains 2 middle level partitions, each middle level partition
contains 365 leaf partitions. Then in this database, we compare the
speed of the old populate SQL (evaluate pg_relation_size on QD via
dispatching, a tuple a dispatch) and the new SQL (a single SQL and only
need to dispatch once). This patch's SQL only takes about 3.8 sec, however,
the old SQL takes about 50 sec.

Authored-by: Miao Chen <miaochen@mail.ustc.edu.cn>
Authored-by: Zhenghua Lyu <kainwen@gmail.com>

kainwen commented:
Since PR pipeline does not run gpexpand, I have built a local dev pipeline to test: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/zlv_test_expand

ashwinstar commented on gpMgmt/bin/gpexpand:
```diff
@@ -1577,19 +1577,60 @@ class gpexpand:
         # including new segments has been started once before so finalize
         self.finalize_prepare()
 
+    def make_populate_sql_with_table_size(self, sql):
```
s/make_populate_sql_with_table_size/generate_sql_to_populate_table_size

kainwen replied:
> s/make_populate_sql_with_table_size/generate_sql_to_populate_table_size

Done. Thanks.

ashwinstar commented on gpMgmt/bin/gpexpand:
```diff
@@ -1577,19 +1577,60 @@ class gpexpand:
+             select oid as table_oid,
+                    pg_relation_size(oid) as size
+             from gp_dist_random('pg_class')
+             where oid >= 16384
```
gpcheckcat defines this `FIRST_NORMAL_OID = 16384`. Seems helpful to define similar for better code reading.

kainwen replied:
> gpcheckcat defines this `FIRST_NORMAL_OID = 16384`. Seems helpful to define similar for better code reading.

Done. Thanks.

ashwinstar commented on gpMgmt/bin/gpexpand:
```diff
@@ -1602,7 +1643,9 @@ WHERE
     AND c.relpersistence != 't'
-                  """ % (undone_status, src_bytes_str)
+                  """ % (undone_status)
+        if not self.options.simple_progress:
```
lets add new line to make it ease for reading code (the indentation of code due to SQL statement makes it harder to follow along)

kainwen replied:
> lets add new line to make it ease for reading code (the indentation of code due to SQL statement makes it harder to follow along)

Done. Also I reformat the SQL. Thanks.

ashwinstar reviewed (commented):
I am guessing there exist tests to provide coverage for the functional correctness of this code change?

Would be helpful if can have some performance numbers to quote in commit message for the change for future reference.

SmartKeyerror commented on gpMgmt/bin/gpexpand:
```diff
@@ -1577,19 +1577,60 @@ class gpexpand:
+        # correctness no matter single stage or multi stage agg.
+        cte_sql = """with table_size_cte(table_oid, size) as
+        (
+           select table_oid, sum(size)
```
It's better to add new tests for this skillful method to get the table size for all tables in `db_size_functions.sql`, to make sure this method is totally correct since I don't know whether the gpexpand pipeline verifies the total size correct or not.

kainwen replied:
> It's better to add new tests for this skillful method to get the table size for all tables in `db_size_functions.sql`, to make sure this method is totally correct since I don't know whether the gpexpand pipeline verifies the total size correct or not.

Good suggestion! I will do it now. Thanks @SmartKeyerror

kainwen replied:
> It's better to add new tests for this skillful method to get the table size for all tables in `db_size_functions.sql`, to make sure this method is totally correct since I don't know whether the gpexpand pipeline verifies the total size correct or not.

Done. @SmartKeyerror Please help review. Thanks!

kainwen commented:
> I am guessing there exist tests to provide coverage for the functional correctness of this code change?
> 
> Would be helpful if can have some performance numbers to quote in commit message for the change for future reference.

Hi @ashwinstar 
 
I have added a guard test case in `subselect_gp` to show we will have the correct plan of the new SQL. cc @SmartKeyerror 

Other parts should be covered by gpexpand behave test suites.

Also I have updated the performance result in commit message:

> We have done local test to show the performance gain. In a local 8*8
> 64-segments cluster, we create 100 partition tables, each partition
> table contains 2 middle level partitions, each middle level partition
> contains 365 leaf partitions. Then in this database, we compare the
> speed of the old populate SQL (evaluate pg_relation_size on QD via
> dispatching, a tuple a dispatch) and the new SQL (a single SQL and only
> need to dispatch once). This patch's SQL only takes about 3.8 sec, however,
> the old SQL takes about 50 sec.

zxuejing commented on gpMgmt/bin/gpexpand:
```diff
@@ -1577,32 +1579,76 @@ class gpexpand:
+                   '%s' as undone_status,
+                   NULL as expansion_started,
+                   NULL as expansion_finished,
+                   0 as source_bytes
```
can we remove source_bytes here?

kainwen replied:
> can we remove source_bytes here?

It is still used for simple progress.

zxuejing commented on gpMgmt/bin/gpexpand:
```diff
@@ -1665,19 +1711,18 @@ WHERE
                 NULL as expansion_started,
                 NULL as expansion_finished,
-                %s as source_bytes
+                0 as source_bytes
```
can we remove source_bytes here?

kainwen replied:
> can we remove source_bytes here?

It is still used in simple progress.

ashwinstar commented on src/test/regress/expected/db_size_functions.out:
```diff
@@ -333,3 +333,31 @@ select pg_relation_size(oid) between 3000000 and 5000000 from pg_class where rel
+create table heapsizetest_size(a bigint);
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
+HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
+copy (select pg_relation_size(oid) from pg_class where relname = 'heapsizetest') to '/tmp/t_heapsizetest_size_xxx';
```
what am I missing when I run this on psql I get 0 rows vs test has output of 1 row

kainwen replied:
> what am I missing when I run this on psql I get 0 rows vs test has output of 1 row

Hi @ashwinstar this case depends on previous SQL to create the table `heapsizetest`.

ashwinstar reviewed (approved):
LGTM

kainwen commented:
This pr only modifies gpexpand related code and add several cases in regress and I have local pipeline with gpexpand job green.

PR pipeline fails due to known issue: one (bitmap_index) is fixed and the other (qp_union_intersect_optimizer) is under fixing.

Push this, thanks for reviewing!
